ops/lib/tcp_load_balancer/tcp_lb.py

```diff
@@ -0,0 +1,241 @@
+            pools = self.tcp_member_pools.pools
+            for pool in self.tcp_member_pools.pools:
+                logger.debug(pools.listener.port)
+                logger.debug(pools.listener.name)
+                for member in pool.members:
+                    logger.debug(member.port)
+                    logger.debug(member.address)
+                    # other fields...
+
+
+`TcpLoadBalancer`_ is the type that exposes the information about members that a TCP
+load-balancer should provide load-balancing for.
+
+    class MyServiceCharm(ops.charm.CharmBase):
+
+        LISTENER_PORT = 80
+        SERVICE_PORT = 8080
+        MONITOR_PORT = 8081
+
+        def __init__(self, framework):
+            super().__init__(framework, None)
+            self.tcp_lb = TcpLoadBalancer(self, 'tcp-lb', lb_algorithm='round_robin')
+            self.framework.observe(self.tcp_lb.on.lb_available, self._on_lb_available)
+
+        def _on_lb_available(self, event):
+            listener = Listener(name=self.app.name.replace('/', '_'), port=self.LISTENER_PORT)
+            fqdn = socket.getfqdn()
+            member = Member(fqdn, self.MEMBER_PORT, monitor_port=self.MONITOR_PORT)
+            health_monitor = HTTPHealthMonitor(timeout=timedelta(seconds=10), http_method='GET',
+                                               url_path='/health?ready=1')
+            self.tcp_lb.expose_member(listener, member, health_monitor)
+            # ...
+'''
+
+
+class PoolsChanged(EventBase):
+
+    '''Event emitted by TcpMemberPools.on.pools_changed.
+
+    This event will be emitted if any of the existing or new members exposes new data over a
+    relation for the load-balancer to re-assess its current state.
+    '''
+    pass
+
+
+class TcpMemberPoolEvents(EventSetBase):
+
+    '''Events emitted by the TcpMemberPoolEvents class.'''
+    pools_changed = EventSource(PoolsChanged)
+
+
+class TcpMemberPools(Object):
+
+    '''Provides the information about pools of TCP load-balancer members.'''
+
+    on = TcpMemberPoolEvents()
+
+    _stored = StoredState()
+
+    def __init__(self, charm, relation_name):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._member_pools = None
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event):
+        self.on.pools_changed.emit()
+
+    @property
+    def pools(self):
+        if self._member_pools is None:
+            self._member_pools = []
+            for relation in self.model.relations[self._relation_name]:
+                if not relation.units:
+                    continue
+                app_data = relation.data[relation.app]
+
+                listener_data = app_data.get('listener')
+                if listener_data is None:
+                    continue
+                listener = pickle.loads(listener_data.encode('utf-8'))
```

Comment:
The encoding step could easily raise an exception. Do you want to handle that possibility?